## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#MLC looking at the data types for missing information.
df.count()


first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [6]:
# What three item categories had the most entries?

three_items_df = df["category"].value_counts()
three_items_df.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?
most_subcate_df = df.subcategory.loc[df.category == "consumables"].value_counts()

most_subcate_df.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
five_clients = df["client_id"].value_counts().head()
five_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [61]:
# Store the client ids of those top 5 clients in a list.
top_clients = list(five_clients.keys())
top_clients

[33615, 66037, 46820, 38378, 24741]

In [62]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client = df.qty.loc[df.client_id == 33615].sum()
top_client

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [1]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df["subtotal"] = df.unit_price * df.qty
df[["unit_price", "qty", "subtotal"]].head()



NameError: name 'df' is not defined

In [64]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def shipping(row):
    cost = 0
    num = row['unit_weight'] * row['qty']
    if num < 50:
        cost = num * 10
    if num > 50:
        cost = num * 7
    return cost
df['shipping_price'] = df.apply(shipping, axis=1)
df[["unit_weight", "qty", "shipping_price"]].head()


unit_weight  qty  shipping_price
0         7.50  105         5512.50
1         1.49   21          312.90
2         1.68   39          458.64
3         1.23   29          356.70
4        46.43   20         6500.20

In [65]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df["line_cost"] = ((df.unit_cost * df.qty) + df.shipping_price).round(2)
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  shipping_price     line_price  line_cost  \
0            1  115164.00         5512.50  131839.076250   85597.05   
1            0     523.95          312.90     914.258625     629.79   
2            6     527.28          458.64    1077.117600     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 23 columns]

In [66]:
# Create a column for the profit of
# each line using line cost and line price

df["line_profit"] = (df.line_price - df.line_cost).round(2)
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number   subtotal  shipping_price     line_price  line_cost  \
0            1  115164.00         5512.50  131839.076250   85597.05   
1            0     523.95          312.90     914.258625     629.79   
2            6     527.28          458.64    1077.117600     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [42]:
# Check your work using the totals 


   

TypeError: tuple indices must be integers or slices, not str

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [86]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
for client in top_clients:
    total = df.loc[df["client_id"] == client, 'line_price'].sum()
    print(f"{client}: ${total.round(2)}")
    
    




33615: $8376762.32
66037: $10259514.8
46820: $9743794.32
38378: $12906550.88
24741: $82268891.98


In [88]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
top_clients_list = []
for client in top_clients:
    qty = df.loc[df["client_id"] == client, 'qty'].sum()
    shipping = df.loc[df["client_id"] == client, 'shipping_price'].sum().round(2)
    line_price = df.loc[df["client_id"] == client, 'line_price'].sum().round(2)
    line_cost = df.loc[df["client_id"] == client, 'line_cost'].sum().round(2)
    line_profit = df.loc[df["client_id"] == client, 'line_profit'].sum().round(2)
    values = {
        'client_id' : client,
        'qty' : qty,
        'shipping_price' : shipping,
        'line_price' : line_price,
        'line_cost' : line_cost,
        'line_profit' : line_profit
    }
    top_clients_list.append(values)

top_client_table_df = pd.DataFrame(top_clients_list)
top_client_table_df.sort_values('line_profit', ascending=False)

                        








client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268891.98  45688899.71  36579992.33
3      38378   73667      3429455.40  12906550.88   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.80   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.32   7007191.64   2736602.69
0      33615   64313      1828484.89   8376762.32   6174813.91   2201948.37

In [94]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
renamed_df = top_client_table_df.rename(columns={"client_id":"Client ID", 
                                          "qty":"Units", 
                                          "shipping_price": "Shipping", 
                                          "line_price": "Total Revenue",
                                           "line_cost": "Total Cost",
                                            "line_profit": "Total Profit"})


millions_of_dollars = renamed_df["Total Profit"]/1000000
renamed_df["Total Profit"] = millions_of_dollars.round(2)
renamed_df["Total Profit"] = "$" + renamed_df["Total Profit"].astype(str) + "M"

millions_of_dollars = renamed_df["Total Cost"]/1000000
renamed_df["Total Cost"] = millions_of_dollars.round(2)
renamed_df["Total Cost"] = "$" + renamed_df["Total Cost"].astype(str) + "M"

millions_of_dollars = renamed_df["Total Revenue"]/1000000
renamed_df["Total Revenue"] = millions_of_dollars.round(2)
renamed_df["Total Revenue"] = "$" + renamed_df["Total Revenue"].astype(str) + "M"

millions_of_dollars = renamed_df["Shipping"]/1000000
renamed_df["Shipping"] = millions_of_dollars.round(2)
renamed_df["Shipping"] = "$" + renamed_df["Shipping"].astype(str) + "M"

renamed_df.head()



Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      33615   64313   $1.83M        $8.38M     $6.17M        $2.2M
1      66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
2      46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M

In [95]:
# Sort the updated data by "Total Profit" form highest to lowest
renamed_df.sort_values('Total Profit', ascending=False)

Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
2      46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.17M        $2.2M